### DISCUSSION TITLE
How to invoke the expansion of LuaSnip snippets programmatically in other .lua files

### DISCUSSION THREAD

**Discussion by kylefhartzenberg** [L3MON4D3/LuaSnip#970] [Q&A] [answered]
### Problem
I have the following snippet defined in the `lua/snippets/tex.lua` file (see the configuration directory structure at the bottom). This snippet successfully creates a LaTeX figure environment whenever `\fg` is pressed inside a `.tex` file. However, I am trying to invoke this snippet programmatically and indirectly via a key-mapping.
```lua
s( -- figure environment
        {
            trig="\\fg",
            dscr="Automatically expand '\\fg' to figure environment",
            snippetType="autosnippet"
        },
        fmt(
            [[
            \begin{figure}
                \centering
                \includegraphics[width=<>\textwidth]{<>}
                \caption{<>}
                \label{fig:<>}
            \end{figure}
            ]],
            { 
                i(2), i(1), i(3), i(4)
            },
            { delimiters = "<>" }
        )
    ),
```

### The Scenario
The (simplified) scenario is as follows:

1. The user presses `Alt + F` within a `.tex` file. This invokes a function called `handle_latex_fig` which is defined in the `lua/texshortcuts.lua` file. The key mapping is set in the `lua/mappings.lua` file.
2. The `handle_latex_fig` function checks if the user is inside a valid LaTeX figure environment via Treesitter.
3. If the user is inside the valid figure environment, additional checks are performed to check the figure's path, whether the file exists, opens the file for editing etc.
4. If the user is not inside a valid figure environment, the function gets the current text under the cursor and then it should invoke the "`\fg`" LaTeX figure environment creation snippet defined in `lua/snippets/tex.lua`, inserting the text that was under the cursor at the first LuaSnip insert node (i.e. as the figure's path). For example:

```tex
\begin{figure}
    \centering
    \includegraphics[width=<>\textwidth]{some_text_that_was_under_cursor}
    \caption{<>}
    \label{fig:<>}
\end{figure}
```

Step 4 is where I am having issues and need help. I have made three attempts at invoking the LaTeX figure environment creation snippet, however, none have produced the desired results. In each of the following examples, "`somefigurename.svg`" is the text under the cursor prior to pressing `Alt + F`.

### Attempted Solutions

**Attempt 1 via Neovim's "feed keys" API using typing mode**
```lua
-- Get text under cursor
local text_under_cursor = get_text_under_cursor()
-- Clear current line
vim.api.nvim_set_current_line("") -- clear current line
-- Invoke LuaSnip figure environment creation snippet
vim.api.nvim_feedkeys("i\\fg", "t", true)
-- Fill in figure include path
vim.api.nvim_feedkeys("i" .. text_under_cursor, "t", true)
```

This clears the current line and replaces it with `\fgsomefigurename.svg`. If only the first `nvim_feedkeys` call is used, the snippet expands successfully. However, whenever both are run together, the keys are all fed sequentially as one piece.

**Attempt 2 via Neovim's "feed keys" API using execute mode**

Modifying the above code slightly by changing [the mode](https://neovim.io/doc/user/builtin.html#feedkeys())  to `x!` forces expansion of the first `nvim_feedkeys` call but requires one to then press the `Esc` key before the second `nvim_feedkeys`is called successfully to input the figure path/name (i.e. the text under the cursor). This attempt works but isn't ideal and feels clunky to use.
```lua
-- Get text under cursor
local text_under_cursor = get_text_under_cursor()
-- Clear current line
vim.api.nvim_set_current_line("") -- clear current line
-- Invoke LuaSnip figure environment creation snippet
vim.api.nvim_feedkeys("i\\fg", "x!", true)
-- Fill in figure include path
vim.api.nvim_feedkeys("a" .. text_under_cursor, "t", true)
```

**Attempt 3 via LuaSnip's "snip expand" API**

```lua
-- Get text under cursor
local text_under_cursor = get_text_under_cursor()
-- Clear current line
vim.api.nvim_set_current_line("") -- clear current line
-- Invoke LuaSnip figure environment creation snippet
local luasnip = require("luasnip")
local snips = luasnip.get_snippets("tex")
luasnip.snip_expand(snips[1]) -- This doesn't actually find the snippet I'm looking for should expand something
-- Fill in figure include path
vim.api.nvim_feedkeys(text_under_cursor, "t", true)
```
This results in an error "`attempt to index local 'snippet' (a nil value)`" which I believe is because the `get_snippets()` function is not executing successfully (i.e. its returning `nil` instead of a table containing the snippets defined for `.tex` files). After spending quite some time reading the documentation, and looking through various open/closed issues (as well as forums) relating to `snip_expand`, I cannot find a example of how to use `snip_expand` (especially in more non-trivial cases like the one described herein). Any guidance would be appreciated.

### Additional Info

I am using the latest version of LuaSnip at the time of writing (commit [e81cbe6](https://github.com/L3MON4D3/LuaSnip/commit/e81cbe6004051c390721d8570a4a0541ceb0df10)) loaded via [lazy.nvim](https://github.com/folke/lazy.nvim). My `.config/nvim` configuration directory structure is as follows. There are no issues to note with the current configuration.
```none
.
├── ftplugin
│   └── html.lua
├── init.lua
├── lua
│   ├── autocmds.lua
│   ├── texshortcuts.lua
│   ├── mappings.lua
│   ├── options.lua
│   ├── package-manager.lua
│   ├── snippets
│   │   ├── all.lua
│   │   ├── html.lua
│   │   └── tex.lua
│   └── templates
│       └── tex-template
└── spell
    ├── en.utf-8.add
    └── en.utf-8.add.spl
```

**Comment by L3MON4D3** [comment]
Attempt 3 sounds promising!
I never really focused on providing some ergonomic way to get a specific snippet by name after it's been added, have you considered storing the snippet somewhere else (ie. in a table like `local named_snippets = {fg = s(...), ... })`) and then getting it from there when adding it in the loader/expanding it manually?

**Reply by kylefhartzenberg** [reply]
Thanks for the quick response! I noticed that so the plan was that if I could get `local snips = luasnip.get_snippets("tex")` to return a table of snippets successfully, I could then iterate over them to check whether the snippet matches the desired snippet. For example (in pseudo-code):
```lua
local luasnip = require("luasnip")
local snips = luasnip.get_snippets("tex")
for _, snip in pairs(snips) do
    if snip.trigger() == "\\fg" then
        luasnip.snip_expand(snip)
    end
end
```
However, there are two issues with this: (1) `luasnip.get_snippets("tex")` or even `luasnip.get_snippets("all")` simply returns nil, and (2) I'm not sure if there is an API to access the trigger or some other unique identifier (like the snippet itself) for a particular snippet to make the `if` comparison. Do you have any suggestions for these two issues? Or, do you think its best to pursue the suggestion of storing the snippet somewhere else?

**Reply by L3MON4D3** [reply]
Ah, the issue with `get_snippets` is probably that the snippets are lazy-loaded, so they most likely don't yet exist when `get_snippets` is called.
I think the best way around this is to define the required snippets somewhere else.
If you want to use `get_snippets`, `snip.trigger` (not a function) contains the trigger (this isn't official, but extremely unlikely to change), and the snippets are loaded on `BufWinEnter`, so if you register an autocommand for that *after* luasnip registered its own (we do it in `plugin/`), that should run after the snippets were loaded.

**Reply by kylefhartzenberg** [reply]
Do you have any recommendations on where/how the snippets should be defined to ensure they exist/are accessible? Currently they are all stored in `~/.config/nvim/lua/snippets` with the relevant file-types as the file-names which are then loaded via LuaSnip's API.

More specifically, my LuaSnip setup and configuration is being called from `~/.config/nvim/lua/package-manager.lua`. In this file, the following actions are performed:
1. Call to `luasnip.config.setup{}`
2. Call to `luasnip.config.set_config({ config_settings_set_here })`
3. Set key mappings for expansion, jumping etc.
4. Load all snippets with a call to `require("luasnip.loaders.from_lua").load({ paths = "~/.config/nvim/lua/snippets/" })`
5. Setup `snip_expand` with `vim.o.ul`
6. Setup and configure [hrsh7th/nvim-cmp](https://github.com/hrsh7th/nvim-cmp) completion engine to work with LuaSnip, LSP, etc.

**Reply by L3MON4D3** [reply]
I think I'd put it into `~/.config/nvim/lua/named-snippets`, like

```lua
-- gives you the same abbreviations available in loaded files
require("luasnip").setup_snip_env()

return {
    fg = s(...),
    othername = s(...),
}
```
and refer to them like `require("named-snippets").fg`

**Reply by kylefhartzenberg** [reply, marked as answer]
@L3MON4D3 Your idea worked perfectly. I added the `named-snippets.lua` file in `.config/nvim/lua` (the same directory as `texshortcuts.lua` and then added the following code to the latter file:
```lua
-- Get text under cursor
local text_under_cursor = get_text_under_cursor()
-- Clear current line
vim.api.nvim_set_current_line("") -- clear current line

local luasnip = require("luasnip")
local fg = require("named-snippets").fg -- Retrieve named snippet
luasnip.snip_expand(fg) -- Invoke LuaSnip figure environment creation snippet
vim.api.nvim_feedkeys(text_under_cursor, "t", true) -- Fill in figure include path 
```

Thank you very much for your help. The work you do on LuaSnip is appreciated.
